### PR TITLE
Do not call GetHashCode() on enums

### PR DESCRIPTION
**Pull request by ladipro**
Calling `GetHashCode()` on an enum results in boxing (i.e. a GC allocation) when running on .NET Framework. This has been addressed in .NET Core (https://github.com/dotnet/coreclr/pull/7895) but we want to fix it on our side still because it's showing in Visual Studio profiles:
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1166715

This PR is also removing the calls to `base.GetHashCode()` which are pointless in structures that calculate the hashcode out of all their fields.

### CODE REVIEW

**Comment by BenVillalobos** [comment]
Closing to reopen and trigger a CI build.

File: src/Build/Logging/EvaluationLocationIdAgnosticComparer.cs [hunk 1]
```diff
@@ -36,13 +36,13 @@ public bool Equals(EvaluationLocation x, EvaluationLocation y)
...
             var hashCode = 1198539463;
-            hashCode = hashCode * -1521134295 + obj.EvaluationPass.GetHashCode();
+            hashCode = hashCode * -1521134295 + EqualityComparer<EvaluationPass>.Default.GetHashCode(obj.EvaluationPass);
             hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(obj.EvaluationPassDescription);
             hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(obj.File);
             hashCode = hashCode * -1521134295 + EqualityComparer<int?>.Default.GetHashCode(obj.Line);
             hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(obj.ElementName);
             hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(obj.ElementDescription);
-            hashCode = hashCode * -1521134295 + obj.Kind.GetHashCode();
+            hashCode = hashCode * -1521134295 + EqualityComparer<EvaluationLocationKind>.Default.GetHashCode(obj.Kind);
```
Comment (Forgind) [inline src/Build/Logging/EvaluationLocationIdAgnosticComparer.cs:45]: I wasn't familiar with this idiom, and when I tried putting it into [sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeALANzFYDMlO5AwuQN7Hm+XNZq5ALIAKAJRcefGQDtyAS3IBecrIB0AMwVwANgBM89IjJlUAnKIBEACwVXxx03wWyY5G9Bt4Vi9QHE4GAAJLzYAe304CSdnSjxLTyhvR2lnV3ckm1ZVAFEARwBXCF0FGABPCLAABwgERAAeWQA+dQAROE0IQt0YAKDQ5Iio0QVUkziLUSyccZkAXzTyaoQFADcIeHI4WUKwNSkJ5209QyRyE4McJcWieaA=), it seemed to do the same thing. What did I do wrong?

Comment (drewnoakes) [reply src/Build/Logging/EvaluationLocationIdAgnosticComparer.cs:45]: `constrained` can box: https://sharplab.io/#v2:EYLgdgpgLgZgHgGiiAlgGwD4AEAMACLARgBYBuAWACgIwBXAWzwFE8BvAXyqIDYCAmPAGE2VPGIIBmAoV4owUPAFlCAChYQAlHgC8APjwQAdAHFoACQCGAZwAWggPYATCCo0VK4ydNnylfNQZaeswAjrQWaChQAJ4O9AAOFgBOEEkAPEy6hgAiEDAWtGhQJubWdk4umu7sQA

**Comment by Forgind** [comment]
Thanks!

File: src/Build/Logging/EvaluationLocationIdAgnosticComparer.cs (hunk 1 above)
Comment (ladipro) [reply src/Build/Logging/EvaluationLocationIdAgnosticComparer.cs:45]: One super hacky way to quickly check if a piece of code allocates is simply running it in a tight `while (true)` loop in a trivial console app and checking the VS diagnostics window.

Seeing small yellow markers appear regularly means that GC is collecting something and that something has to be allocations performed in the loop.

image

**Pull request merged**